repo-scripts/size-analysis/analysis-helper.ts

```diff
@@ -0,0 +1,316 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import * as tmp from 'tmp';
+import * as path from 'path';
+import * as fs from 'fs';
+import * as rollup from 'rollup';
+import * as terser from 'terser';
+import * as ts from 'typescript';
+import resolve from 'rollup-plugin-node-resolve';
+import commonjs from 'rollup-plugin-commonjs';
+
+/** Contains a list of members by type. */
+export type MemberList = {
+  classes: string[];
+  functions: string[];
+  variables: string[];
+  enums: string[];
+};
+/** Contains the dependencies and the size of their code for a single export. */
+export type ExportData = { dependencies: MemberList; sizeInBytes: number };
+
+/**
+ * This functions builds a simple JS app that only depends on the provided
+ * export. It then uses Rollup to gather all top-level classes and functions
+ * that that the export depends on.
+ *
+ * @param exportName The name of the export to verify
+ * @param jsBundle The file name of the source bundle that contains the export
+ * @return A list of dependencies for the given export
+ */
+export async function extractDependencies(
+  exportName: string,
+  jsBundle: string,
+  map: Map<string, string>
+): Promise<MemberList> {
+  const { dependencies } = await extractDependenciesAndSize(
+    exportName,
+    jsBundle,
+    map
+  );
+  return dependencies;
+}
+
+/**
+ * Helper for extractDependencies that extracts the dependencies and the size
+ * of the minified build.
+ */
+export async function extractDependenciesAndSize(
+  exportName: string,
+  jsBundle: string,
+  map: Map<string, string>
+): Promise<ExportData> {
+  const input = tmp.fileSync().name + '.js';
+  const output = tmp.fileSync().name + '.js';
+
+  const beforeContent = `export { ${exportName} } from '${path.resolve(
+    jsBundle
+  )}';`;
+
+  fs.writeFileSync(input, beforeContent);
+
+  // Run Rollup on the JavaScript above to produce a tree-shaken build
+  const bundle = await rollup.rollup({
+    input,
+    plugins: [
+      resolve({
+        mainFields: ['esm2017', 'module', 'main']
+      }),
+      commonjs()
+    ]
+  });
+  await bundle.write({ file: output, format: 'es' });
+
+  const dependencies = extractDeclarations(output, map);
+
+  // Extract size of minified build
+  const afterContent = fs.readFileSync(output, 'utf-8');
+
+  const { code } = terser.minify(afterContent, {
+    output: {
+      comments: false
+    },
+    mangle: true,
+    compress: false
+  });
+
+  fs.unlinkSync(input);
+  fs.unlinkSync(output);
+  return { dependencies, sizeInBytes: Buffer.byteLength(code!, 'utf-8') };
+}
+
+/**
+ * Extracts all function, class and variable declarations using the TypeScript
+ * compiler API.
+ */
+export function extractDeclarations(
+  jsFile: string,
+  map?: Map<string, string>
```

Comment:
`map` is only available for analyzing js bundle, right?
Can you also comment on what the map is, and under what condition it will be conditional?